Lib/gftools/packager.py

```diff
@@ -0,0 +1,1312 @@
+  """
+      getting this as a shallow copy, because for some files we want to
+      search in the filesystem.
+
+      branch_or_tag: as used in `git clone -b`
+
+      NOTE: libgit2 and hence pygit2 doesn't support shallow clones yet,
+      but that's the most lightweight way to get the whole directory
+      structure.
+  """
+
+  return subprocess.run(['git', 'clone', '--depth', '1', '--bare'
+                       , '-b', branch_or_tag, git_url
+                       , target_dir], check=True
+                       , stdout=subprocess.PIPE)
+
+def _shallow_clone_github(target_dir, gh_repo_name_with_owner, branch_or_tag='master'):
+  """
+      Another way to get this data would be the github api, but there's
+      a quota (see response headers "X-Ratelimit-Limit: 60" and
+      "X-Ratelimit-Remaining"
+      https://developer.github.com/v3/#rate-limiting
+      unauthorized requests (per IP): the rate limit allows for up to 60 requests per hour.
+      Basic Authentication or OAuth: you can make up to 5000 requests per hour.
+  """
+  git_url = GITHUB_REPO_HTTPS_URL(gh_repo_name_with_owner=gh_repo_name_with_owner)
+  return _shallow_clone_git(target_dir, git_url, branch_or_tag=branch_or_tag)
+
+CATEGORIES = ['DISPLAY', 'SERIF', 'SANS_SERIF', 'SANS_SERIF',
+                  'HANDWRITING', 'MONOSPACE']
+
+upstream_yaml_schema = Map({
+    'name': Str(),
+    'repository_url': Str(), # TODO: custom validation please
+    'branch': Str(),
+    'category': Enum(CATEGORIES),
+    'designer': Str(),
+    # allowing EmptyDict here, even though we need files in here,
+    # but we will catch missing files later in the process.
+    # When we have repository_url and branch we can add a editor based
+    # dialog that suggests all files present in the repo (delete lines of
+    # files you don't want to include).
+    'files': EmptyDict() | MapPattern(Str(), Str()) # Mappings with arbitrary key names
+})
+
+# since upstream_yaml_template is incomplete, it can't be parsed with
+# the complete upstream_yaml_schema. Here's a more forgiving schema for
+# the template.
+upstream_yaml_template_schema = Map({
+    'name': EmptyNone() | Str(),
+    'repository_url': EmptyNone() | Str(), # TODO: custom validation please
+    'branch': EmptyNone() | Str(),
+    'category':  EmptyNone() | Enum(CATEGORIES),
+    'designer': EmptyNone() |Str(),
+    'files': EmptyDict() | MapPattern(Str(), Str())
+})
+
+upstream_yaml_template  = f'''
+# Please edit this upstream configuration for the family accordingly.
+# This is a yaml formatted file.
+# An "#" (number sign) denotes a comment.
+# For more help see the docs at:
+# https://github.com/googlefonts/gf-docs/tree/master/METADATA
+
+# Full family name, with initial upper cases and spaces
+name:
+
+# In most cases this should be based on the GitHub https repo url:
+# this https://github.com/{{owner}}/{{repo}}.git
+repository_url:
+
+# The branch name used to update google fonts. e.g.: master
+branch:
```

Comment:
It would be nice to have `master` as a placeholder. Just ran the tool and forgot to fill this and it crashed.

Imo, I would prefer us to only push fonts to production that are in the master branch.

Reply:
master as default is good for me.

> Imo, I would prefer us to only push fonts to production that are in the master branch.

I see alternative branches as way around of having to put font binaries into the repo. That way I can make a bad practices "gf" branch, lol.

Actually, more imortant: for a QA workflow, in a feature branch, before publishing, it's good to be able to use a branch. It's not only about pushing to production.